// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: app_backend_api/app/v1/query.proto

package appv1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on ConfigRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ConfigRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ConfigRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ConfigRequestMultiError, or
// nil if none found.
func (m *ConfigRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ConfigRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ConfigRequestMultiError(errors)
	}

	return nil
}

// ConfigRequestMultiError is an error wrapping multiple validation errors
// returned by ConfigRequest.ValidateAll() if the designated constraints
// aren't met.
type ConfigRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ConfigRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ConfigRequestMultiError) AllErrors() []error { return m }

// ConfigRequestValidationError is the validation error returned by
// ConfigRequest.Validate if the designated constraints aren't met.
type ConfigRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ConfigRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ConfigRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ConfigRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ConfigRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ConfigRequestValidationError) ErrorName() string { return "ConfigRequestValidationError" }

// Error satisfies the builtin error interface
func (e ConfigRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sConfigRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ConfigRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ConfigRequestValidationError{}

// Validate checks the field values on ConfigResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ConfigResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ConfigResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ConfigResponseMultiError,
// or nil if none found.
func (m *ConfigResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ConfigResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(m.GetItems()) > 0 {

		for idx, item := range m.GetItems() {
			_, _ = idx, item

			if all {
				switch v := interface{}(item).(type) {
				case interface{ ValidateAll() error }:
					if err := v.ValidateAll(); err != nil {
						errors = append(errors, ConfigResponseValidationError{
							field:  fmt.Sprintf("Items[%v]", idx),
							reason: "embedded message failed validation",
							cause:  err,
						})
					}
				case interface{ Validate() error }:
					if err := v.Validate(); err != nil {
						errors = append(errors, ConfigResponseValidationError{
							field:  fmt.Sprintf("Items[%v]", idx),
							reason: "embedded message failed validation",
							cause:  err,
						})
					}
				}
			} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
				if err := v.Validate(); err != nil {
					return ConfigResponseValidationError{
						field:  fmt.Sprintf("Items[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					}
				}
			}

		}

	}

	if len(errors) > 0 {
		return ConfigResponseMultiError(errors)
	}

	return nil
}

// ConfigResponseMultiError is an error wrapping multiple validation errors
// returned by ConfigResponse.ValidateAll() if the designated constraints
// aren't met.
type ConfigResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ConfigResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ConfigResponseMultiError) AllErrors() []error { return m }

// ConfigResponseValidationError is the validation error returned by
// ConfigResponse.Validate if the designated constraints aren't met.
type ConfigResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ConfigResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ConfigResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ConfigResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ConfigResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ConfigResponseValidationError) ErrorName() string { return "ConfigResponseValidationError" }

// Error satisfies the builtin error interface
func (e ConfigResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sConfigResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ConfigResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ConfigResponseValidationError{}

// Validate checks the field values on ConfigItem with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ConfigItem) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ConfigItem with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in ConfigItemMultiError, or
// nil if none found.
func (m *ConfigItem) ValidateAll() error {
	return m.validate(true)
}

func (m *ConfigItem) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ConfigItemMultiError(errors)
	}

	return nil
}

// ConfigItemMultiError is an error wrapping multiple validation errors
// returned by ConfigItem.ValidateAll() if the designated constraints aren't met.
type ConfigItemMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ConfigItemMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ConfigItemMultiError) AllErrors() []error { return m }

// ConfigItemValidationError is the validation error returned by
// ConfigItem.Validate if the designated constraints aren't met.
type ConfigItemValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ConfigItemValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ConfigItemValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ConfigItemValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ConfigItemValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ConfigItemValidationError) ErrorName() string { return "ConfigItemValidationError" }

// Error satisfies the builtin error interface
func (e ConfigItemValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sConfigItem.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ConfigItemValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ConfigItemValidationError{}

// Validate checks the field values on AppIdRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *AppIdRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on AppIdRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in AppIdRequestMultiError, or
// nil if none found.
func (m *AppIdRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *AppIdRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if utf8.RuneCountInString(m.GetId()) != 24 {
		err := AppIdRequestValidationError{
			field:  "Id",
			reason: "value length must be 24 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)

	}

	if len(errors) > 0 {
		return AppIdRequestMultiError(errors)
	}

	return nil
}

// AppIdRequestMultiError is an error wrapping multiple validation errors
// returned by AppIdRequest.ValidateAll() if the designated constraints aren't met.
type AppIdRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m AppIdRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m AppIdRequestMultiError) AllErrors() []error { return m }

// AppIdRequestValidationError is the validation error returned by
// AppIdRequest.Validate if the designated constraints aren't met.
type AppIdRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e AppIdRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e AppIdRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e AppIdRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e AppIdRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e AppIdRequestValidationError) ErrorName() string { return "AppIdRequestValidationError" }

// Error satisfies the builtin error interface
func (e AppIdRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sAppIdRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = AppIdRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = AppIdRequestValidationError{}

// Validate checks the field values on AppIdsRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *AppIdsRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on AppIdsRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in AppIdsRequestMultiError, or
// nil if none found.
func (m *AppIdsRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *AppIdsRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for WithDeleted

	if len(errors) > 0 {
		return AppIdsRequestMultiError(errors)
	}

	return nil
}

// AppIdsRequestMultiError is an error wrapping multiple validation errors
// returned by AppIdsRequest.ValidateAll() if the designated constraints
// aren't met.
type AppIdsRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m AppIdsRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m AppIdsRequestMultiError) AllErrors() []error { return m }

// AppIdsRequestValidationError is the validation error returned by
// AppIdsRequest.Validate if the designated constraints aren't met.
type AppIdsRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e AppIdsRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e AppIdsRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e AppIdsRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e AppIdsRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e AppIdsRequestValidationError) ErrorName() string { return "AppIdsRequestValidationError" }

// Error satisfies the builtin error interface
func (e AppIdsRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sAppIdsRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = AppIdsRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = AppIdsRequestValidationError{}
